docs/TICCL-indexer.1
.TH TICCL-indexer 1 "2020 mar 09"

.SH NAME
TICCL-indexer - a tool to create an exhaustive index to all lexical variant pairs
given a particular Levenshtein or edit distance in a corpus or the corpus focus list.

.SH SYNOPSIS

TICCL-indexer [options]

TICCL-indexerNT [options]

.SH DESCRIPTION
.B TICCL-indexer
and
.B TICCL-indexerNT
both process an anagram file (as generated by
.B TICCL-anahash
) and a character confusion file (as generated by
.B FoLiA-lexstat
) and create a possibly very large file of character confusion anagram values
chained to the lower word anagram values for word pairs displaying the
particular character confusion.

.B TICCL-indexer
is recommended for large corpora, i.e. when you have more items in the corpus
frequency list than you have in the character confusion file. It can run on multiple threads, at
the cost of a higher memory usage.

.B TICCL-indexerNT
is recommended for relatively small corpora. It can run on multiple threads, at
the cost of a higher memory usage.

.SH OPTIONS
.B --hash
anagramfile
.RS
the file generated by
.B TICCL-anahash
.
.RE

.B --charconf
character confusionfile
.RS
the character confusion file generated by
.B TICCL-lexstat
.
.RE

.B --foci
focifile
.RS
Use 'focifile' to select only the anagram hash values actually observed in the corpus itself 
with a frequency lower than the artifrq to work on. This file is generated by
.B TICCL-anahash
when the
.B --artifrq
option is used. This limits the overall search space and the amount of work to be done.
.RE

.B -o
outputfile
.RS
the name of the output file. Whatever you choose here, the extension will be set
to '.index' for
.B TICCL-indexer
and to '.indexNT' for
.B TICCL-indexerNT
.RE

.B --confstats
statsfile
.RS
Use 'statsfile' to create a separate list of observed confusion statistics. (optional)
.RE


.B --low
low
.RS
let the program exclude entries from the anagram file that are shorter than 'low' characters.
The default is 5.
.RE

.B --high
high
.RS
let the program exclude entries from the anagram file that are longer than 'high' characters.
The default is 35.
.RE

.B -t
or
.B --threads
num_threads
.RS
use 'num_threads' threads to run on. You may us --threads="max" to use as many
threads as possible. This will allocate 2 processors less than given by the
$OMP_NUM_THREADS environment variable, leaving some processor power for other
purposes.
.RE

.B -v
.RS
be more verbose.
.RE

.B -V
or
.B --version
.RS
Show VERSION
.RE

.B -h
or
.B --help
.RS
this message
.RE

.SH BUGS
possibly

.SH AUTHORS
Ko van der Sloot lamasoftware@science.ru.nl

Martin Reynaert reynaert@uvt.nl

.SH SEE ALSO
.BR TICCL-lexstat (1)
.BR TICCL-anahash (1)
.BR FoLiA-stats (1)


NEWS
ticcltools 0.6 2018-06-05
[Ko vander Sloot]
Intermediate release, with a lot of new code to handle N-grams
Also a lot of refactoring is done, for more clear and maintainable code.
This is work in progress still.

* TICCL-unk:
  - more extensive acronym detection
  - fixed artifreq problems in 'clean' punctuated words
  - added filters for 'unwanted' characters
  - added a ligature filter to convert evil ligatures
  - normalize all hyphens to a 'normal' one (-)
  - use a better definition of punctuation (unicode character class is not
    good enough to decide)

* TICCL-lexstat:
  - the 'separator' symbol should get freq=0, so it isn't counted
  - the clip value is added to the output filename

* TICCL-indexer:
  - indexer and indexerNT now produce the same output, using different
    strategies when a --foci files is used.

* TICCL-LDcalc:
  major overhaul for n-grams
  - added a ngram point column to the output (so NOT backward compatible!)
  - produce a '.short' list for short word corrections
  - produce a '.ambi' file with a list of n-grams related to short words
  - prune a lot of ngrams from the output

* TICCL-rank:
 - output is sorted now
 - honor the ngram-points from the new LDcalc. (so NOT backward compatible!)

* TICCL-chain: new module to chain ranked files

* TICCL-lexclean:
  -added a -x option for 'inverse' alphabet

* TICCL-anahash:
  - added a --list option to produce a list of words and anagram values

[Maarten van Gompel]
* added metadata file: codemeta.json

ticcltools 0.5 2018-02-19
[Ko van der Sloot]
* updated configuration. also for Mac OSX
* use of more ticcutils stuff: diacritics filter
* added a TICCL-mergelex program
* the OMP_THREAD_LIMIT environment variable was ignored sometimes
* TICCL-unk:
   - fixed a problem in artifreq handling
   - changed acronym detection (work in  progress)
   - added -o option
  TICCL-lexstat:
   - added TTR output
   - added -o option
  TICCL-indexer
  - now also handles --foci file. with some speed-up
  - added a -t option
  TICCL-LDcalc:
  - be less picky on a few wrong lines in the data
* added some tests
* when libroaring is installed we built roaring versions of some modules (experimental)
* updated man pages

ticcltools 0.4 2017-04-04
[Ko van der Sloot]
* first official release.
  - added functions to test on Word2Vec datafiles
  - refactoring and modernizing stuff all around

ticcltools 0.3 2016-01-21
* upload ready
ticcltools 0.2 2016-01-14
[Ko van der Sloot]
* repository moved to GitHub
* Travis support
* first 'distributable' version
* added TICCL-stats program
* added W2V-near and W2V-dist programs bases on new libticcl lib


[0.1] Ko vd Sloot
 started autoconfiscating


src/TICCL-indexer.cxx
/*
  Copyright (c) 2006 - 2020
  CLST  - Radboud University
  ILK   - Tilburg University

  This file is part of ticcltools

  ticcltools is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 3 of the License, or
  (at your option) any later version.

  ticcltools is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, see <http://www.gnu.org/licenses/>.

  For questions and suggestions, see:
      https://github.com/LanguageMachines/ticcltools/issues
  or send mail to:
      lamasoftware (at ) science.ru.nl

*/
#include <cassert>
#include <unistd.h>
#include <set>
#include <map>
#include <limits>
#include <algorithm>
#include <vector>
#include <climits>
#include <cstdlib>
#include <string>
#include <stdexcept>
#include <iostream>
#include <sstream>
#include <fstream>
#include "ticcutils/StringOps.h"
#include "ticcutils/CommandLine.h"
#include "ticcutils/Unicode.h"
#include "ticcl/ticcl_common.h"

#include "config.h"
#ifdef HAVE_OPENMP
#include "omp.h"
#endif

using namespace std;
using namespace icu;

void usage( const string& name ){
  cerr << name << endl;
  cerr << "options: " << endl;
  cerr << "\t--hash=<anahash>\tname of the anagram hashfile. (produced by TICCL-anahash)" << endl;
  cerr << "\t--charconf=<charconf>\tname of the character confusion file. (produced by TICCL-lexstat)" << endl;
  cerr << "\t--foci=<focifile>\tname of the file produced by the --artifrq parameter of TICCL-anahash." << endl;
  cerr << "\t\t\t This file is used to limit the searchspace" << endl;
  cerr << "\t-o <outputfile>\t\tname for the outputfile. " << endl;
  cerr << "\t--confstats=<statsfile>\tcreate a list of confusion statistics"
       << endl;
  cerr << "\t--low=<low>\t skip entries from the anagram file shorter than "
       << endl;
  cerr << "\t\t\t'low' characters. (default = 5)" << endl;
  cerr << "\t--high=<high>\t skip entries from the anagram file longer than "
       << endl;
  cerr << "\t\t\t'high' characters. (default=35)" << endl;
  cerr << "\t-t <threads> or --threads <threads>\n\t\t\t Number of threads to run on." << endl;
  cerr << "\t\t\t If 'threads' has the value \"max\", the number of threads is set to a" << endl;
  cerr << "\t\t\t reasonable value. ($OMP_NUM_TREADS - 2)" << endl;
  cerr << "\t-v\t\t run verbose " << endl;
  cerr << "\t-V or --version\t show version " << endl;
  cerr << "\t-h or --help\t this message " << endl;
}

struct experiment {
  set<bitType>::const_iterator start;
  set<bitType>::const_iterator finish;
};


void handle_confs( const experiment& exp,
		   size_t& count,
		   const set<bitType>& anaSet,
		   const set<bitType>& focSet,
		   ostream &of,
		   ostream *csf ){
  bitType vorige = 0;
  bitType totalShift = 0;
  auto sit = exp.start;
  while ( sit != exp.finish ){
    set<bitType> result;
#pragma omp critical(count)
    {
      if ( ++count % 100 == 0 ){
	cout << ".";
	cout.flush();
	if ( count % 5000 == 0 ){
	  cout << endl << count << endl;
	}
      }
    }
    bitType confusie = *sit;
    bitType diff = confusie - vorige;
    totalShift += diff;
    auto it1 = anaSet.begin();
    auto it2 = anaSet.begin();
    while ( it1 != anaSet.end() && it2 != anaSet.end() ){
      bitType v1 = *it1;
      bitType v2 = *it2;
      bitType v2_save = v2;
      if ( v2 >= totalShift ) {
	v2 -= totalShift;
      }
      else {
	v2 = 0;
      }
      if ( v1 == v2 ){
	bool foc = true;
	if ( !focSet.empty() ){
	  // do we have to focus?
	  foc = !( focSet.find( v1 ) == focSet.end()
		   && focSet.find( v2_save ) == focSet.end() );
	  // not if both values out of focus
	}
	if ( foc ){
	  result.insert(v1);
	}
	++it1;
	++it2;
      }
      else if ( v1 < v2 ){
	++it1;
      }
      else {
	++it2;
      }
    }
    vorige = confusie;
    ++sit;
    if ( !result.empty() ){
      stringstream ss;
      ss << confusie << "#";
      set<bitType>::const_iterator it = result.begin();
      while ( it != result.end() ){
	ss << *it;
	++it;
	if ( it != result.end() ){
	  ss << ",";
	}
      }
#pragma omp critical(update)
      {
	if ( csf ){
	  *csf << confusie << "#" << result.size() << endl;
	}
	of << ss.str() << endl;
      }
    }
  }
}

size_t init( vector<experiment>& exps,
	     const set<bitType>& hashes,
	     size_t threads ){
  exps.clear();
  size_t partsize = hashes.size() / threads;
  if ( partsize < 1 ){
    experiment e;
    e.start = hashes.begin();
    e.finish = hashes.end();
    exps.push_back( e );
    return 1;
  }
  set<bitType>::const_iterator s = hashes.begin();
  for ( size_t i=0; i < threads; ++i ){
    experiment e;
    e.start = s;
    for ( size_t j=0; j < partsize && s != hashes.end(); ++j ){
      ++s;
    }
    e.finish = s;
    exps.push_back( e );
  }
  if ( s != hashes.end() ){
    exps[exps.size()-1].finish = hashes.end();
  }
  return threads;
}

int main( int argc, char **argv ){
  TiCC::CL_Options opts;
  try {
    opts.set_short_options( "vVho:t:" );
    opts.set_long_options( "charconf:,hash:,low:,high:,help,version,"
			   "foci:,threads:,confstats:" );
    opts.init( argc, argv );
  }
  catch( TiCC::OptionError& e ){
    cerr << e.what() << endl;
    usage( argv[0] );
    exit( EXIT_FAILURE );
  }
  string progname = opts.prog_name();
  if ( opts.extract('h') || opts.extract("help") ){
    usage( progname );
    exit(EXIT_SUCCESS);
  }
  if ( opts.extract('V') || opts.extract("version") ){
    cerr << PACKAGE_STRING << endl;
    exit(EXIT_SUCCESS);
  }
  if ( argc < 3	){
    usage( progname );
    exit(EXIT_FAILURE);
  }
  bool verbose = opts.extract( 'v' );
  string anahashFile;
  string confFile;
  string fociFile;
  string outFile;
  string confstats_file;
  int lowValue = 5;
  int highValue = 35;
  opts.extract( "hash", anahashFile );
  opts.extract( "charconf", confFile );
  opts.extract( "confstats", confstats_file );
  opts.extract( "foci", fociFile );
  opts.extract( 'o', outFile );
  string value;
  if ( opts.extract("low", value ) ){
    if ( !TiCC::stringTo(value,lowValue) ) {
      cerr << "illegal value for --low (" << value << ")" << endl;
      exit( EXIT_FAILURE );
    }
  }
  if ( opts.extract("high", value ) ){
    if ( !TiCC::stringTo(value,highValue) ) {
      cerr << "illegal value for --high (" << value << ")" << endl;
      exit( EXIT_FAILURE );
    }
  }
  int numThreads=1;
  value = "1";
  if ( !opts.extract( 't', value ) ){
    opts.extract( "threads", value );
  }
#ifdef HAVE_OPENMP
  if ( TiCC::lowercase(value) == "max" ){
    numThreads = omp_get_max_threads() - 2;
  }
  else {
    if ( !TiCC::stringTo(value,numThreads) ) {
      cerr << "illegal value for -t (" << value << ")" << endl;
      exit( EXIT_FAILURE );
    }
  }
#else
  if ( value != "1" ){
    cerr << "unable to set number of threads!.\nNo OpenMP support available!"
	 <<endl;
    exit(EXIT_FAILURE);
  }
#endif
  if ( !opts.empty() ){
    cerr << "unsupported options : " << opts.toString() << endl;
    usage(progname);
    exit(EXIT_FAILURE);
  }
  ifstream ana( anahashFile );
  if ( !ana ){
    cerr << "problem opening corpus anagram hashfile: " << anahashFile << endl;
    exit(1);
  }
  ifstream conf( confFile );
  if ( !conf ){
    cerr << "problem opening charconfusion file: " << confFile << endl;
    exit(1);
  }

  set<bitType> focSet;
  if ( !fociFile.empty() ){
    ifstream foc( fociFile );
    if ( !foc ){
      cerr << "problem opening foci file: " << fociFile << endl;
      exit(1);
    }
    while ( foc ){
      bitType bit;
      foc >> bit;
      foc.ignore( INT_MAX, '\n' );
      focSet.insert( bit );
    }
    cout << "read " << focSet.size() << " foci values" << endl;
  }

  if ( outFile.empty() ){
    outFile = anahashFile;
    string::size_type pos = outFile.rfind(".");
    if ( pos != string::npos ){
      outFile = outFile.substr(0,pos);
    }
    outFile += ".index";
  }
  else if ( !TiCC::match_back( outFile, ".index" ) ){
    outFile += ".index";
  }

  ofstream of( outFile );
  if ( !of ){
    cerr << "problem opening outputfile: " << outFile << endl;
    exit(1);
  }
  ofstream *csf = 0;
  if ( !confstats_file.empty() ){
    csf = new ofstream( confstats_file );
    if ( !csf ){
      cerr << "problem opening outputfile: " << confstats_file << endl;
      exit(1);
    }
  }
  cout << "reading corpus word anagram hash values" << endl;
  size_t skipped = 0;
  set<bitType> anaSet;
  string line;
  while ( getline( ana, line ) ){
    vector<string> parts;
    if ( TiCC::split_at( line, parts, "~" ) > 1 ){
      bitType bit = TiCC::stringTo<bitType>( parts[0] );
      vector<string> parts2;
      if ( TiCC::split_at( parts[1], parts2, "#" ) > 0 ){
	UnicodeString firstItem = TiCC::UnicodeFromUTF8( parts2[0] );
	if ( firstItem.length() >= lowValue &&
	     firstItem.length() <= highValue ){
	  anaSet.insert( bit );
	}
	else {
	  if ( verbose ){
	    cerr << "skip " << parts2[0] << endl;
	  }
	  ++skipped;
	}
      }
    }
  }
  cout << "read " << anaSet.size() << " corpus anagram values" << endl;
  cout << "skipped " << skipped << " out-of-band corpus anagram values" << endl;

  cout << "reading character confusion anagram values" << endl;
  set<bitType> confSet;
  size_t count = 0;
  while ( getline( conf, line ) ){
    vector<string> parts;
    if ( ++count % 1000 == 0 ){
      cout << ".";
      cout.flush();
      if ( count % 50000 == 0 ){
	cout << endl << count << endl;;
      }
    }
    if ( TiCC::split_at( line, parts, "#" ) > 0 ){
      bitType bit = TiCC::stringTo<bitType>( parts[0] );
      confSet.insert( bit );
    }
    else {
      cerr << "problems with line " << line << endl;
      cerr << "bail out " << endl;
      exit(1);
    }
  }
  cout << endl << "read " << confSet.size()
       << " character confusion anagram values" << endl;

  vector<experiment> experiments;
  size_t expsize = init( experiments, confSet, numThreads );
#ifdef HAVE_OPENMP
  omp_set_num_threads( expsize );
  cout << "running on " << expsize << " threads." << endl;
#endif


  cout << "processing all character confusion values" << endl;
  count = 0;
#pragma omp parallel for shared( experiments, of, csf )
  for ( size_t i=0; i < expsize; ++i ){
    handle_confs( experiments[i], count, anaSet, focSet, of, csf );
  }
  cout << "\nwrote indexes into: " << outFile << endl;
  if ( csf ){
    cout << "wrote confusion statistics into: " << confstats_file << endl;
    csf->close();
    delete csf;
  }
}
